### PR TITLE
[Examples] Add PIC example

### PR DESCRIPTION
**Pull request by ylchan87**
Rewrite the PIC example in new API.
Tested to work with following additional args in conf file:
input=/user/ylchan/testPICdata.txt
dimension=2
maxIter=1
stopThres=-1.0

### CODE REVIEW

**Comment by lijinf2** [comment]
Hi Yat-long, thanks a lot for your PR. Before we merge your PR, can you present some experimental data? For example, can you show the running time of the new implementation as compared to the old one? You may refer to https://bitbucket.org/ddmbr/husky/pull-requests/162/examples-pic-optimization/diff

**Comment by kygx-legend** [comment]
Please rebase at first

**Comment by ylchan87** [comment]
I just rebased, not sure if done right.

There are a lot of difference that make comparision with the old.
But here's the runtime test:
Setting: 4 worker , 2 proc each. 1000 2D datapoint. 1 iteration. Output to screen, not HDFS
Time: 3.468s / 3.795s / 3.235s

File: examples/pic.cpp [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    }
+    globalize(velem_list);
+
+    husky::base::log_msg( "vec "
+                          + std::to_string(velem_list.get_size()) + " "
+                          + std::to_string(startPos) + " "
+                          + std::to_string(endPos) + " "
+                        );
+
+    //FIXME:: Instead of the whole object, should push only the val and id of VElem to the Mat
```
Comment (ddmbr) [inline examples/pic.cpp]: Usually it's `FIXME(name)`, in which the `name` refers to person who should handle it. @kygx-legend

File: examples/pic.cpp [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+    husky::ObjList<DimStat> rowstat_list;
+    auto& rowsumCh =
+        husky::ChannelFactory::create_push_combined_channel<float, husky::SumCombiner<float>>(affMatElem_list, rowstat_list);
+
+    list_execute(affMatElem_list, [&](MatElem & e){
+        auto pts = pairCh.get(e);
+        float nSigmaSq = 0;
+
+        //if(e.id()==10011){
```
Comment (ddmbr) [inline examples/pic.cpp]: You may just remove these lines

File: examples/pic.cpp [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+        husky::ChannelFactory::create_push_channel<NDpoint>(ndpoint_list, affMatElem_list);
+
+    //FIXME:: seems stupid to create a Channel just to make a list accessable by all worker?
+    auto& dimstatBCh = husky::ChannelFactory::create_broadcast_channel<int,DimStat>(dimstat_list);
+    list_execute(dimstat_list, [&](DimStat& d) {
+        dimstatBCh.broadcast( d.id(), d);
+    });
+
+    ////Why this doesn't work? I get vec of size 0 even dimstat_list is globalized
+    //std::vector<DimStat> local_dimstat_list = dimstat_list.get_data();
```
Comment (ddmbr) [inline examples/pic.cpp]: You may discuss with us but don't leave them in the code. By the way, you are copying a std::vector.

File: examples/pic.cpp [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    }
+
+    KeyT pos;
+    float val;
+};
+
+class VElem {
+   //Element for the classification vector V
+   public:
+    //using KeyT = std::pair<uint, uint>; // std::hash don't like std::pair
```
Comment (ddmbr) [inline examples/pic.cpp]: Remove the comment

File: examples/pic.cpp [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+        while (it != tok.end()) {
+            coords.push_back(stod(*it++));
+        }
+        ndpoint_list.add_object(NDpoint(id, coords));
+    };
+    load(infmt, parse_ndpoint);
+    globalize(ndpoint_list);
+    husky::base::log_msg( "ndpoint_list size " + std::to_string( ndpoint_list.get_size()));
+
+    husky::base::log_msg( "part1 ok ");
```
Comment (ddmbr) [inline examples/pic.cpp]: I wonder whether it's possible to make the logs more meaningful? E.g., something like "Data points loaded". And I guess they will flood the screen if you use many threads?

**Comment by ddmbr** [comment]
Ping @ylchan87 . Also please format your code according to [this guide](https://github.com/husky-team/husky/blob/master/CONTRIBUTING.md).

**Comment by ylchan87** [comment]
you mean the commit msg?

2016-10-27 11:35 GMT+08:00 YANG Fan <notifications@github.com>:

**Comment by kygx-legend** [comment]
It means the coding style including naming must be consistent with the project.

**Pull request closed without merging**